### PR TITLE
Implement resource boolean expression parser

### PR DESCRIPTION
**Pull request by kimlisa**
part of [rfd 55](https://github.com/gravitational/teleport/blob/master/rfd/0055-webui-ss-paginate-filter.md#filter-query-language) on resource query language

#### Description
Created a predicate parser that's customized for our `ResourceWithLabels` (rules and examples defined in [rfd](https://github.com/gravitational/teleport/blob/master/rfd/0055-webui-ss-paginate-filter.md#filter-query-language)). Plan is to allow UI and CLI users to use this predicate language to filter resources by labels and json fields.

Also updates `vulcand/predicate` to latest `v1.2.0`

### CODE REVIEW

**Comment by kimlisa** [comment]
friendly ping  :pray:  @espadolini @rosstimothy

File: lib/services/parser.go
Comment (espadolini) [inline lib/services/parser.go:605]: Isn't it be weird for `labels.foo.bar.baz` to match against `combinedLabels["foo"]` while disregarding `bar` and `baz`? Do we want to avoid errors here for some reason? Same goes for `name.foo`.

**Comment by espadolini** [comment]
Is there a reason why we're planning on switching to our own fork of `github.com/vulcand/predicate`? gravitational/predicate#1 seems to be merged in `vulcand/predicate:master` already.

File: lib/services/parser.go
Comment (rosstimothy) [inline lib/services/parser.go:595]: What if `len(fields) <= 0`?

File: lib/services/parser.go
Comment (rosstimothy) [inline lib/services/parser.go:605]: What if `len(fields) < 1` ?

File: lib/services/parser.go
Comment (rosstimothy) [inline lib/services/parser.go:616]: What if `len(fields) <= 1`?

File: lib/services/parser.go
Comment (rosstimothy) [inline lib/services/parser.go:570]: Whats the difference between `equals` and `EQ`?

Comment (rosstimothy) [reply lib/services/parser.go:570]: But are they equivalent? Is `(a == b)` the same as `equals(a, b)`?

 If so can this just directly call `EQ` instead of checking anything?

File: lib/services/parser.go
Comment (espadolini) [inline lib/services/parser.go:564]: Should this be `true` instead? Or perhaps the whole `NEQ` could be made into a negation of the function in `EQ`?

**Comment by kimlisa** [comment]
@espadolini 
> Is there a reason why we're planning on switching to our own fork of github.com/vulcand/predicate? 

it was under sasha's request, he has plans to extend it further (and is the original creator of the library) 

> gravitational/predicate#1 seems to be merged in vulcand/predicate:master already.

i originally made a PR against the actual repo, and folks from another organization got to reviewing it and merged it before sasha could view it, so sasha wanted it to go another round of review with someone from our team. I plan to just close that PR (since it's already in)

File: lib/services/parser.go
Comment (kimlisa) [reply lib/services/parser.go:595]: this will not happen b/c no fields and no operand are errors, i updated test to reflect more edge cases

File: lib/services/parser.go
Comment (kimlisa) [reply lib/services/parser.go:605]: same as above

File: lib/services/parser.go
Comment (kimlisa) [reply lib/services/parser.go:616]: same as above

File: lib/services/parser.go
Comment (kimlisa) [reply lib/services/parser.go:570]: @rosstimothy is right, i refactored it to use the same methods for `eq, neq, and equals`

File: lib/services/parser.go
Comment (rosstimothy) [reply lib/services/parser.go:571]: This is a guarantee of the predicate library?

File: lib/services/parser.go
Comment (rosstimothy) [inline lib/services/parser.go:551]:
```suggestion
			EQ:  isOperandsEqual,
```

File: lib/services/parser.go
Comment (rosstimothy) [inline lib/services/parser.go:555]:
```suggestion
			"equals": isOperandsEqual,
```

File: lib/services/parser.go
Comment (rosstimothy) [inline lib/services/parser.go:629]:
```suggestion
func areOperandsEqual(a interface{}, b interface{}) bool {
```

File: lib/services/parser.go
Comment (kimlisa) [reply lib/services/parser.go:551]: `eqPred` is necessary as it returns a closure which is expected by the parser

File: lib/services/parser.go
Comment (kimlisa) [inline lib/services/parser.go:632]: this is similar to https://github.com/gravitational/predicate/blob/master/lib.go#L57, i am wondering if it's more proper to do this in the predicate library: create a `areOperandsEqual` and have `Equals` return a closure calling `areOperandsEqual`. And then here, we can extend the switch in `areOperandsEqual` with a label case.

File: lib/services/parser.go
Comment (kimlisa) [reply lib/services/parser.go:571]: yes, it goes through like switch statements to build expressions, and returns errors as it encounters misfits

File: lib/services/parser.go
Comment (rosstimothy) [reply lib/services/parser.go:632]: I'm not so sure that should belong in predicate. I think we can instead rely on `pedicate.Equals` here. What do you think about this change?
```suggestion
func areOperandsEqual(a interface{}, b interface{}) bool {
	switch aval := a.(type) {
	case label:
		bval, ok := b.(string)
		return ok && aval.value == bval
	default:
		return predicate.Equals(a, b)()
	}
}
```

File: lib/services/parser.go
Comment (espadolini) [inline lib/services/parser.go:597]:
```suggestion
				if len(fields) > 1 {
					return nil, trace.BadParameter("only one field are supported with identifier %q, got %d: %v", ResourceNameIdentifier, len(fields), fields)
				}
				return resource.GetName(), nil
```

**Review by espadolini — APPROVED** [review]
LGTM other than one style nit

File: lib/services/parser.go
Comment (kimlisa) [reply lib/services/parser.go:632]: i think it's legit, ty!

Comment (espadolini) [reply lib/services/parser.go:629]: Perhaps make `areOperandsEqual` return a `predicate.BoolPredicate`?

Comment (kimlisa) [reply lib/services/parser.go:629]: @espadolini I needed it to return a bool so I could use it flexibly like below. 
```go
        eqPred := func(a interface{}, b interface{}) predicate.BoolPredicate {
		return func() bool {
			return areOperandsEqual(a, b)
		}
	}
	neqPred := func(a interface{}, b interface{}) predicate.BoolPredicate {
		return func() bool {
			return !areOperandsEqual(a, b)
		}
	}
```

Or we could use same technique, if we want it to return a `predicate.BoolPredicate`:
```go
        eqPred := areOperandsEqual // don't need it in a variable anymore but for example sake
	neqPred := func(a interface{}, b interface{}) predicate.BoolPredicate {
		return func() bool {
			return !areOperandsEqual(a, b)()
		}
	}
```
Is this what you were thinking of?

Comment (espadolini) [reply lib/services/parser.go:629]:
```go
neqPred := areOperandsEqual.Not()
```

would work, thankfully it's from `v1.1.0` which is what we're using already.

Comment (kimlisa) [reply lib/services/parser.go]: ty @espadolini https://github.com/gravitational/teleport/pull/10008/commits/539f700fe0530b8ddd791ece330a69155da9e46a

**Review by rosstimothy — APPROVED** [review]
LGTM, any idea why the tests are failing?

**Comment by kimlisa** [comment]
@rosstimothy it's b/c i haven't updated the predicate library that had a recent commit to it that allowed additional syntax, i am waiting for this approval before I rebase: https://github.com/gravitational/teleport/pull/10110

**Comment by espadolini** [comment]
Why not just bump the version as part of this PR?

**Comment by kimlisa** [comment]
@espadolini  i thought i needed to make it a separate PR, but i included it in here

File: go.mod
Comment (espadolini) [inline go.mod:221]: Bump the dependency on `github.com/vulcand/predicate` to v1.2.0 as well, it's still at v1.1.0

Comment (espadolini) [reply go.mod:221]: Huh, I guess we actually don't need to do that; I don't know what the best practices are, do we just leave whatever version was there at the time of replacement?

Comment (espadolini) [reply go.mod:221]: Actually, considering that `v1.2.0` is identical for both, can't we just keep using `github.com/vulcand/predicate` for now and change the actual module name for `v1.3.0`, as it gets developed further? We only have 5 imports in here, we can just fix those as we swap the package

Comment (kimlisa) [reply go.mod]: @espadolini i did feel silly about both repos being same version, i did as you suggested, just upgrade vulcand

**Review by rudream — APPROVED** [review]

**Review by alex-kovoy — APPROVED** [review]

**Pull request merged**
